### PR TITLE
COMPRESS-644 -- do not allow zero by entries when detecting magic-less tar

### PR DESCRIPTION
**Pull request by tballison**
When relying on reading the first record from a tar file and then verifying the checksum as proof that the file is a tar file, require that the record have more than zero bytes.

### CODE REVIEW

File: src/test/java/org/apache/commons/compress/DetectArchiverTestCase.java [hunk 1]
```diff
@@ -141,4 +145,17 @@ public void testEmptyJarArchive() throws Exception {
...
+
+    @Test
+    public void ignoreZeroByteEntryInTarDetect_COMPRESS644() throws IOException {
+        try (final InputStream in =
+                     new BufferedInputStream(Files.newInputStream(
+                             Paths.get("src/test/resources/org/apache/commons" +
+                             "/compress/COMPRESS-644/ARW05UP.ICO")))) {
+            ArchiveStreamFactory.detect(in);
+            fail("should have thrown ArchiveException");
+        } catch (ArchiveException e) {
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/compress/DetectArchiverTestCase.java:157]: Use assertThrows().

File: src/main/java/org/apache/commons/compress/archivers/ArchiveStreamFactory.java [hunk 2]
```diff
@@ -268,7 +269,16 @@ public static String detect(final InputStream in) throws ArchiveException {
...
+                // COMPRESS-644 - do not allow zero byte file entries
+                TarArchiveEntry tae = tais.getNextTarEntry();
+                //try to find the first non-directory entry within the first 10 entries.
+                int max = 10;
+                int count = 0;
+                while (tae != null && tae.isDirectory() && count++ < max) {
+                    tae = tais.getNextTarEntry();
+                }
+                if (tae != null && !tae.isDirectory()
+                        && tae.getSize() > 0l && tae.isCheckSumOK()) {
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/compress/archivers/ArchiveStreamFactory.java:281]: 01?

File: src/main/java/org/apache/commons/compress/archivers/ArchiveStreamFactory.java [hunk 3]
```diff
@@ -268,7 +269,16 @@ public static String detect(final InputStream in) throws ArchiveException {
             try {
                 tais = new TarArchiveInputStream(new ByteArrayInputStream(tarHeader));
                 // COMPRESS-191 - verify the header checksum
-                if (tais.getNextTarEntry().isCheckSumOK()) {
+                // COMPRESS-644 - do not allow zero byte file entries
+                TarArchiveEntry tae = tais.getNextTarEntry();
+                //try to find the first non-directory entry within the first 10 entries.
+                int max = 10;
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/compress/archivers/ArchiveStreamFactory.java:275]: What makes 10 special? Needs a comment or a better a Javadoc comment for a new constant.

**Review by garydgregory — CHANGES_REQUESTED** [review]
@tballison 
Thank you for the PR. Please see my comments.

File: src/main/java/org/apache/commons/compress/archivers/ArchiveStreamFactory.java (hunk 3 above)
Comment (tballison) [reply src/main/java/org/apache/commons/compress/archivers/ArchiveStreamFactory.java]: Completely arbitrary.  I improved documentation.  I'm completely open to the idea of dropping this attempt to iterate through the first x directory entries.

File: src/main/java/org/apache/commons/compress/archivers/ArchiveStreamFactory.java [hunk 4]
```diff
@@ -268,7 +278,15 @@ public static String detect(final InputStream in) throws ArchiveException {
             try {
                 tais = new TarArchiveInputStream(new ByteArrayInputStream(tarHeader));
                 // COMPRESS-191 - verify the header checksum
-                if (tais.getNextTarEntry().isCheckSumOK()) {
+                // COMPRESS-644 - do not allow zero byte file entries
+                TarArchiveEntry tae = tais.getNextTarEntry();
+                //try to find the first non-directory entry within the first 10 entries.
+                int count = 0;
+                while (tae != null && tae.isDirectory() && count++ < MAX_TAR_RECORDS) {
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/compress/archivers/ArchiveStreamFactory.java:285]: This causes detection to fail for `src/test/resources/COMPRESS-612/test-times-star-folder.tar` (once you rebase on git master which tests more files).

**Review by garydgregory — COMMENTED** [review]
Hello @tballison 
Thanks for your patience.
The PR breaks detecting old format tar archives in an improved test since this PR was created. 
I tweaked the detection from this PR and pushed a different fix in git master.

**Comment by tballison** [comment]
It looks like the changes here were accepted. I'm closing out this PR. Thank you @garydgregory and team!

**Pull request closed without merging**